sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java

```diff
@@ -0,0 +1,225 @@
+  }
+
+  @VisibleForTesting
+  static class BatchingDoFn<K, InputT> extends DoFn<KV<K, InputT>, KV<K, Iterable<InputT>>> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BatchingDoFn.class);
+    private static final String END_OF_WINDOW_ID = "endOFWindow";
+    private static final String BATCH_ID = "batch";
+    private static final String NUM_ELEMENTS_IN_BATCH_ID = "numElementsInBatch";
+    private static final String KEY_ID = "key";
+    private final long batchSize;
+    private final Duration allowedLateness;
+
+    @TimerId(END_OF_WINDOW_ID)
+    private final TimerSpec timer = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+    @StateId(BATCH_ID)
+    private final StateSpec<Object, BagState<InputT>> batchSpec;
+
+    @StateId(NUM_ELEMENTS_IN_BATCH_ID)
+    private final StateSpec<Object, AccumulatorCombiningState<Long, Long, Long>>
+        numElementsInBatchSpec;
+
+    @StateId(KEY_ID)
+    private final StateSpec<Object, ValueState<K>> keySpec;
+
+    private final long prefetchFrequency;
+
+    BatchingDoFn(
+        long batchSize,
+        Duration allowedLateness,
+        Coder<K> inputKeyCoder,
+        Coder<InputT> inputValueCoder) {
+      this.batchSize = batchSize;
+      this.allowedLateness = allowedLateness;
+      this.batchSpec = StateSpecs.bag(inputValueCoder);
+      this.numElementsInBatchSpec =
+          StateSpecs.combiningValue(
+              VarLongCoder.of(),
+              new Combine.CombineFn<Long, Long, Long>() {
+
+                @Override
+                public Long createAccumulator() {
+                  return 0L;
+                }
+
+                @Override
+                public Long addInput(Long accumulator, Long input) {
+                  return accumulator + input;
+                }
+
+                @Override
+                public Long mergeAccumulators(Iterable<Long> accumulators) {
+                  long sum = 0L;
+                  for (Long accumulator : accumulators) {
+                    sum += accumulator;
+                  }
+                  return sum;
+                }
+
+                @Override
+                public Long extractOutput(Long accumulator) {
+                  return accumulator;
+                }
+              });
+
+      this.keySpec = StateSpecs.value(inputKeyCoder);
+      // prefetch every 20% of batchSize elements. Do not prefetch if batchSize is too little
+      this.prefetchFrequency = ((batchSize / 5) <= 1) ? Long.MAX_VALUE : (batchSize / 5);
+    }
+
+    @ProcessElement
+    public void processElement(
+        @TimerId(END_OF_WINDOW_ID) Timer timer,
+        @StateId(BATCH_ID) BagState<InputT> batch,
+        @StateId(NUM_ELEMENTS_IN_BATCH_ID)
+            AccumulatorCombiningState<Long, Long, Long> numElementsInBatch,
+        @StateId(KEY_ID) ValueState<K> key,
+        ProcessContext c,
+        BoundedWindow window) {
+      Instant firingInstant = window.maxTimestamp().plus(allowedLateness);
```

Comment:
yes, it is exactly for window expiration. I coded it that way until `@OnWindowExpiration` is available.